runtime/vm/jfr.cpp

```diff
@@ -804,6 +823,90 @@ jfrExecutionSampleCallback(J9VMThread *currentThread, IDATA handlerKey, void *us
+{
+	PORT_ACCESS_FROM_VMC(currentThread);
+	OMRPORT_ACCESS_FROM_J9PORT(PORTLIB);
+
+	omrthread_process_time_t currentProcCPUTimes = {0};
+	intptr_t processTimeRC = omrthread_get_process_times(&currentProcCPUTimes);
+
+	J9SysinfoCPUTime currentSysCPUTime = {0};
+	intptr_t sysTimeRC = omrsysinfo_get_CPU_utilization(&currentSysCPUTime);
+
+	if ((0 == processTimeRC) && (0 == sysTimeRC)) {
+		J9JFRCPULoad *jfrEvent = (J9JFRCPULoad *)reserveBuffer(currentThread, sizeof(J9JFRCPULoad));
+		if (NULL != jfrEvent) {
+			initializeEventFields(currentThread, (J9JFREvent *)jfrEvent, J9JFR_EVENT_TYPE_CPU_LOAD);
+
+			JFRState *jfrState = &currentThread->javaVM->jfrState;
+			uintptr_t numberOfCpus = j9sysinfo_get_number_CPUs_by_type(J9PORT_CPU_ONLINE);
+			int64_t currentTime = j9time_nano_time();
+			omrthread_process_time_t *prevProcCPUTimes = &jfrState->prevProcCPUTimes;
+			int64_t *prevProcTimestamp = &jfrState->prevProcTimestamp;
+
+			if (-1 == *prevProcTimestamp) {
+				jfrEvent->jvmUser = 0;
+				jfrEvent->jvmSystem = 0;
+			} else {
+				int64_t timeDelta = currentTime - *prevProcTimestamp;
+				jfrEvent->jvmUser = OMR_MIN((currentProcCPUTimes._userTime - prevProcCPUTimes->_userTime) / ((double)numberOfCpus * timeDelta), 1.0);
+				jfrEvent->jvmSystem = OMR_MIN((currentProcCPUTimes._systemTime - prevProcCPUTimes->_systemTime) / ((double)numberOfCpus * timeDelta), 1.0);
+			}
+			*prevProcCPUTimes = currentProcCPUTimes;
+			*prevProcTimestamp = currentTime;
+
+			J9SysinfoCPUTime *prevSysCPUTime = &jfrState->prevSysCPUTime;
+			if (-1 == prevSysCPUTime->timestamp) {
+				jfrEvent->machineTotal = 0;
+			} else {
+				jfrEvent->machineTotal = OMR_MIN((currentSysCPUTime.cpuTime - prevSysCPUTime->cpuTime) / ((double)numberOfCpus * (currentSysCPUTime.timestamp - prevSysCPUTime->timestamp)), 1.0);
+			}
+			*prevSysCPUTime = currentSysCPUTime;
+		}
+	}
+}
+
+void
+jfrThreadCPULoad(J9VMThread *currentThread, J9VMThread *sampleThread)
+{
+	PORT_ACCESS_FROM_VMC(currentThread);
+
+	omrthread_thread_time_t threadTime;
+
+	intptr_t rc = omrthread_get_thread_times(&threadTime);
+
+	if (-1 != rc) {
+		J9JFRThreadCPULoad *jfrEvent = (J9JFRThreadCPULoad *)reserveBuffer(currentThread, sizeof(*jfrEvent));
+		if (NULL != jfrEvent) {
+			initializeEventFields(currentThread, (J9JFREvent *)jfrEvent, J9JFR_EVENT_TYPE_THREAD_CPU_LOAD);
+
+			ThreadJFRState *jfrState = &sampleThread->threadJfrState;
+			int64_t currentTime = j9time_nano_time();
+			int64_t *prevTimestamp = &jfrState->prevTimestamp;
+			omrthread_thread_time_t *prevThreadCPUTimes = &jfrState->prevThreadCPUTimes;
+
+			if (-1 == *prevTimestamp) {
+				jfrEvent->user = 0;
+				jfrEvent->system = 0;
+			} else {
+				int64_t timeDelta = currentTime - *prevTimestamp;
+				jfrEvent->user = OMR_MIN((threadTime.userTime - prevThreadCPUTimes->userTime) / ((double)timeDelta), 1.0);
+				jfrEvent->system = OMR_MIN((threadTime.sysTime - prevThreadCPUTimes->sysTime) / ((double)timeDelta), 1.0);
```

Comment:
The parentheses in `((double)timeDelta)` are redundant; please remove them (they couldn't possibly alter the meaning).